enterprise/server/scim/scim.go

```diff
@@ -418,57 +418,109 @@ func mapRole(ur *UserResource) (role.Role, error) {
+	userRole, err := mapRole(&ur)
+	if err != nil {
+		return err
+	}
+	u.SubID = saml.SubIDForUserName(ur.UserName, g)
+	u.FirstName = ur.Name.GivenName
+	u.LastName = ur.Name.FamilyName
+	u.Email = ur.UserName
+	u.Groups = []*tables.GroupRole{{Group: *g, Role: uint32(userRole)}}
+	return nil
 }
 
 func (s *SCIMServer) createUser(ctx context.Context, r *http.Request, g *tables.Group) (interface{}, error) {
 	req, err := io.ReadAll(r.Body)
 	if err != nil {
 		return nil, err
 	}
-	log.CtxDebugf(ctx, "SCIM create user request: %s", string(req))
+	log.CtxInfof(ctx, "SCIM create user request:\n%s", string(req))
 	ur := UserResource{}
 	if err := json.Unmarshal(req, &ur); err != nil {
 		return nil, err
 	}
 
-	pk, err := tables.PrimaryKeyForTable("Users")
-	if err != nil {
-		return nil, err
-	}
 	userRole, err := mapRole(&ur)
 	if err != nil {
 		return nil, err
 	}
-	roleUpdate, err := roleUpdateRequest(pk, userRole)
-	if err != nil {
+
+	// If the user exists in our system, but is not currently part of the group
+	// managed by SCIM then all the SCIM query APIs will not return information
+	// about this user.
+	// From the perspective of the upstream system the user effectively does not
+	// exist in our system. If we get a create request for such a user then
+	// we need to translate it to an update request.
+	updateExistingUser := false
+	user, err := s.env.GetAuthDB().LookupUserFromSubID(ctx, saml.SubIDForUserName(ur.UserName, g))
+	if err != nil && !status.IsNotFoundError(err) {
 		return nil, err
 	}
-	u := &tables.User{
-		UserID:    pk,
-		SubID:     saml.SubIDForUserName(ur.UserName, g),
-		FirstName: ur.Name.GivenName,
-		LastName:  ur.Name.FamilyName,
-		Email:     ur.UserName,
-		Groups: []*tables.GroupRole{
-			{Group: *g, Role: uint32(userRole)},
-		},
+	if err == nil {
+		// User already exists in our system.
+
+		// User shouldn't already be part of the SCIM-managed group. If it is,
+		// the upstream system should be sending update requests not create
+		// requests.
+		for _, eg := range user.Groups {
+			if eg.GroupID == g.GroupID {
+				return status.AlreadyExistsErrorf("user %q already exists", ur.UserName), nil
+			}
+		}
+		updateExistingUser = true
+	} else {
+		// User doesn't exist. We can do a normal create.
+
+		pk, err := tables.PrimaryKeyForTable("Users")
+		if err != nil {
+			return nil, err
+		}
+		user = &tables.User{UserID: pk}
 	}
-	if err := s.env.GetUserDB().InsertUser(ctx, u); err != nil {
+
+	if err := fillUserFromResource(user, ur, g); err != nil {
 		return nil, err
 	}
-	if err := s.env.GetUserDB().UpdateGroupUsers(ctx, g.GroupID, roleUpdate); err != nil {
+
+	roleUpdate, err := roleUpdateRequest(user.UserID, userRole, updateExistingUser)
+	if err != nil {
 		return nil, err
 	}
-	return newUserResource(u, g)
+	if updateExistingUser {
+		// Need to add the user to the group first in order to get past the
+		// permission check in UpdateUser.
```

Comment:
nit: could you clarify which permission check this is working around?

Reply:
Improved comment.